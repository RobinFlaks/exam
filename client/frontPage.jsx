import {Link} from "react-router-dom";
import * as React from "react";

export function FrontPage() {

    return <div>
        <h2>hello!</h2>
        <div>
            <Link to={{pathname: "login"}}>login?</Link>
        </div>
    </div>
}