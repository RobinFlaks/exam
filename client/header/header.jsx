import {Link} from "react-router-dom";


export function Header(){
    return(<div>
        <h4>Header:</h4>
        <Link to={{pathname: ""}}>back to front page</Link>
        <div>
        <Link to={{pathname: "login"}}>Login?</Link> or <Link to={{pathname: "register"}}>Register?</Link>
        </div>
        <h4>Header end</h4>
    </div>)
}