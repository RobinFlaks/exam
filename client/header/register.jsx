import React, {useState} from "react";
import { useNavigate } from "react-router-dom";


function FormInput({value, label, onChangeValue}){
  return(
      <div>
        <label>
          {label}:{" "}
          <input value={value} onChange={(e) => onChangeValue(e.target.value)}/>
        </label>
      </div>
  )
}

export function Register({userApi}) {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const navigate = useNavigate();

  async function handleSubmit(e) {
    e.preventDefault();

    userApi.registerUser({username, password});

    setUsername("");
    setPassword("");

    navigate("/");
  }

  return (
    <div>
      <h3>register:</h3>
      <form onSubmit={handleSubmit}>
        <FormInput label={"username"} value={username} onChangeValue={setUsername}/>
        <FormInput label={"password"} value={password} onChangeValue={setPassword}/>
        <button>Register</button>
      </form>
    </div>
  );
}
