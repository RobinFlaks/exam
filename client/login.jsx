import * as React from "react";

export function Login(){
    return(
        <div>
            <h3>this is the Login page :D</h3>
        </div>
    )
}