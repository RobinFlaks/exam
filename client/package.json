{
  "name": "client",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "build": "parcel build index.html",
    "dev": "parcel watch index.html",
    "test": "jest"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@babel/preset-env": "^7.20.2",
    "@babel/preset-react": "^7.18.6",
    "babel-jest": "^29.3.1",
    "jest": "^29.3.1",
    "jest-environment-jsdom": "^29.3.1",
    "jsdom": "^20.0.3",
    "parcel": "^2.8.0",
    "process": "^0.11.10"
  },
  "dependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-router-dom": "^6.4.3"
  },
  "babel": {
    "presets": [
      "@babel/preset-env",
      "@babel/preset-react"
    ]
  },
  "jest": {
    "testEnvironment": "jsdom",
    "globals": {
      "IS_REACT_ACT_ENVIRONMENT": true
    },
    "collectCoverageFrom": [
      "*.jsx"
    ]
  }
}
