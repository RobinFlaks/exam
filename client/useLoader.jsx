import { useEffect, useState } from "react";

export async function fetchJSON(url, options = {}) {
  const res = await fetch(url, {
    method: options.method || "get",
    headers: options.json ? { "content-type": "application/json" } : {},
    body: options.json && JSON.stringify(options.json),
  });

  if (!res.ok) {
    throw new Error(`loading error: ${res.status} -> ${res.statusText};`);
  }

  if (res.status === 200) {
    return await res.json();
  }
}

export function useLoader(loadingFn) {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState();
  const [data, setData] = useState();

  useEffect(() => {
    const loadData = async () => {
      setLoading(true);
      try {
        setData(await loadingFn());
      } catch (error) {
        setError(error);
      }
    };
    loadData().then((r) => setLoading(false));
  }, []);

  return { loading, error, data };
}
