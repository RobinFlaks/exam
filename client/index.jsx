import * as React from "react";
import { createRoot } from "react-dom/client";
import { BrowserRouter, Route, Routes } from "react-router-dom";
import { FrontPage } from "./site/frontPage";
import { Login } from "./header/login";
import { Register } from "./header/register";
import { Menu } from "./site/menu";
import { Header } from "./header/header";
import {fetchJSON} from "./useLoader";

const element = document.getElementById("app");

const root = createRoot(element);

function Application() {
    const menuApi = {
        async listMenu(){
            return fetchJSON("api/general/menu");
        }
    }
    const userApi = {
        async returnUser(){
            return fetchJSON("api/general/users");
        },

        async registerUser(user){
            fetch("/api/general/users", {
                method: "post",
                body: JSON.stringify(user),
                headers:{
                    "content-type": "application/json",
                },
            });
        }
    }
  return (
    <BrowserRouter>
      <Header />
      <Routes>
        <Route path="/" element={<FrontPage userApi={userApi}/>} />
        <Route path="/login" element={<Login />} />
        <Route path="/register" element={<Register userApi={userApi}/>} />
        <Route path="/menu" element={<Menu menuApi={menuApi}/>} />
      </Routes>
    </BrowserRouter>
  );
}

root.render(<Application />);
