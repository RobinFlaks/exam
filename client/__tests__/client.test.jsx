import * as React from "react";
import {createRoot } from "react-dom/client";
import {act} from "react-dom/test-utils";
import {Login} from "../header/login";

describe("client test suite", () =>{

    it("frontpage is served", () =>{
        const element = document.createElement("div");

        const root = createRoot(element);

        act (() => {
            root.render(<Login/>)
        });

        expect(element.querySelector("h3")?.innerHTML).toEqual("this is the Login page :D")
        expect(element.innerHTML).toMatchSnapshot();
    });
});