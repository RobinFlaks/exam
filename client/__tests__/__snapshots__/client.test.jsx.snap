// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client test suite frontpage is served 1`] = `"<div><h3>this is the Login page :D</h3></div>"`;
